From: Haiyue Wang <haiyuewa@163.com>
Subject: xz: fix the preboot code to match the kmalloc_obj change
Date: Sun, 22 Feb 2026 20:11:00 +0800

kmalloc was replaced with kmalloc_obj in lib/xz.  Make the same change
also in the preboot code to fix building of XZ-compressed kernel on a few
archs.  For example, on ARM:

In file included from arch/arm/boot/compressed/../../../../lib/decompress_unxz.c:241,
                 from arch/arm/boot/compressed/decompress.c:56:
arch/arm/boot/compressed/../../../../lib/xz/xz_dec_stream.c: In function 'xz_dec_init':
arch/arm/boot/compressed/../../../../lib/xz/xz_dec_stream.c:787:28: error: implicit declaration of function 'kmalloc_obj'; did you mean 'kmalloc'? [-Wimplicit-function-declaration]
   787 |         struct xz_dec *s = kmalloc_obj(*s);
       |                            ^~~~~~~~~~~
       |                            kmalloc
arch/arm/boot/compressed/../../../../lib/xz/xz_dec_stream.c:787:28: error: initialization of 'struct xz_dec *' from 'int' makes pointer from integer without a cast [-Wint-conversion]
arch/arm/boot/compressed/../../../../lib/decompress_unxz.c:242:
arch/arm/boot/compressed/../../../../lib/xz/xz_dec_lzma2.c: In function 'xz_dec_lzma2_create':
arch/arm/boot/compressed/../../../../lib/xz/xz_dec_lzma2.c:1141:34: error: initialization of 'struct xz_dec_lzma2 *' from 'int' makes pointer from integer without a cast [-Wint-conversion]
  1141 |         struct xz_dec_lzma2 *s = kmalloc_obj(*s);
       |                                  ^~~~~~~~~~~
arch/arm/boot/compressed/../../../../lib/decompress_unxz.c:243:
arch/arm/boot/compressed/../../../../lib/xz/xz_dec_bcj.c: In function 'xz_dec_bcj_create':
arch/arm/boot/compressed/../../../../lib/xz/xz_dec_bcj.c:594:32: error: initialization of 'struct xz_dec_bcj *' from 'int' makes pointer from integer without a cast [-Wint-conversion]
   594 |         struct xz_dec_bcj *s = kmalloc_obj(*s);
       |                                ^~~~~~~~~~~
CC      arch/arm/boot/compressed/fdt.o

[lasse.collin@tukaani.org: changelog updates]
Link: https://lkml.kernel.org/r/20260222121107.116775-1-haiyuewa@163.com
Fixes: 69050f8d6d07 ("treewide: Replace kmalloc with kmalloc_obj for non-scalar types")
Fixes: bf4afc53b77a ("Convert 'alloc_obj' family to use the new default GFP_KERNEL argument")
Signed-off-by: Haiyue Wang <haiyuewa@163.com>
Acked-by: Lasse Collin <lasse.collin@tukaani.org>
Reviewed-by: Kees Cook <kees@kernel.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 lib/decompress_unxz.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/lib/decompress_unxz.c~xz-fix-arm-fdt-compile-error-for-kmalloc-replacement
+++ a/lib/decompress_unxz.c
@@ -157,11 +157,11 @@
  * when XZ_DYNALLOC is used, but the pre-boot free() doesn't support it.
  * Workaround it here because the other decompressors don't need it.
  */
-#undef kmalloc
+#undef kmalloc_obj
 #undef kfree
 #undef vmalloc
 #undef vfree
-#define kmalloc(size, flags) malloc(size)
+#define kmalloc_obj(type) malloc(sizeof(type))
 #define kfree(ptr) free(ptr)
 #define vmalloc(size) malloc(size)
 #define vfree(ptr) do { if (ptr != NULL) free(ptr); } while (0)
_
